airbyte-integrations/connectors/source-shortio/source_shortio/source.py

```diff
@@ -0,0 +1,266 @@
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+
+
+from abc import ABC
+from typing import Any, Iterable, List, Mapping, MutableMapping, Optional, Tuple
+
+import requests
+import json
+import datetime
+from airbyte_cdk.sources import AbstractSource
+from airbyte_cdk.sources.streams import Stream
+from airbyte_cdk.sources.streams.http import HttpStream
+from airbyte_cdk.sources.streams.http.auth import TokenAuthenticator
+from airbyte_cdk.logger import AirbyteLogger
+
+class BasicAuthenticator(TokenAuthenticator):
+
+    def get_auth_header(self) -> Mapping[str, Any]:
+        if self.auth_method is None:
+            return {self.auth_header: f"{self._token}"}
+        else:
+            return {self.auth_header: f"{self.auth_method} {self._token}"}
+
+
+class Links(HttpStream, ABC):
+
+    url_base = "https://api.short.io/api"
+    limit = 150
+    primary_key = "id"
+    before_id = None
+    def next_page_token(
+        self, 
+        response: requests.Response
+        ) -> Optional[Mapping[str, Any]]:
+        
+        links = json.loads(response.text)['links']
+        try:
+            id_string = sorted(links, key=lambda k: k['createdAt'], reverse=False)[0]['idString']
+            if self.before_id != id_string:
+                self.before_id = id_string        
+                return id_string
+            else:
+                return None
+        except IndexError:
+            return None
+    def request_params(
+        self,
+        stream_state: Mapping[str, Any] = None,
+        stream_slice: Mapping[str, Any] = None,
+        next_page_token: Mapping[str, Any] = None
+    ) -> MutableMapping[str, Any]:
+        return {
+            "limit" : self.limit,
+            "domain_id" : self._config['domain_id'],
+            "before": next_page_token or None
+        }
+
+    def path(
+        self, 
+        stream_state: Mapping[str, Any] = None, 
+        stream_slice: Mapping[str, Any] = None,
+        next_page_token: Mapping[str, Any] = None
+    ) -> str:
+        # Get all the links 
+        return "/links"
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        """
+        The short.io API can be incosistent in its inclusion of UTM parameters.
+        Here we specifically check if they've been provided and in case they haven't attempt to fetch from the URL directly.
```

Comment:
```suggestion
        Here, we check if they've been provided and if they haven't, attempt to extract it from the original url.
```

Reply:
Does this mean there are instances where the UTM params exists in the URL but short.io doesn't return it?

Reply:
btw what does UTM stand for?

Reply:
@davinchia correct. Even though their docs reference that they should be present we haven't had them returned consistently. I believe the behaviour depends based on how the link is originally created (whether through their UI or through the API etc. The most common use case for short links is attribution operations (e.g. sharing a link with influencers and including parameters in the original URL in order to trace back purchases and website visits). 

Details on UTM parameters [here](https://en.wikipedia.org/wiki/UTM_parameters)